src/client/opamClient.ml

```diff
@@ -635,6 +635,150 @@ let init_checks ?(hard_fail_exn=true) init_config =
   if hard_fail && hard_fail_exn then OpamStd.Sys.exit_because `Configuration_error
   else not (soft_fail || hard_fail)
 
+let windows_checks config =
+  let vars = OpamFile.Config.global_variables config in
+  let env =
+    List.map (fun (v, c, s) -> v, (lazy (Some c), s)) vars
+    |> OpamVariable.Map.of_list
+  in
+  let success cygcheck =
+    let config =
+      let os_distribution = OpamVariable.of_string "os-distribution" in
+      let update vars =
+        OpamFile.Config.with_global_variables
+          ((os_distribution, S "cygwin", "Set by opam init")::vars)
+          config
+      in
+      match OpamStd.List.pick (fun (v,_,_) ->
+          OpamVariable.equal v os_distribution)
+          vars with
+      | Some (_, S "cygwin", _), _ -> config
+      | None, vars ->  update vars
+      | Some (_, vc, _), vars ->
+        OpamConsole.warning
+          "'os-distribution' already set to another value %s"
+          (OpamVariable.string_of_variable_contents vc);
+        if OpamConsole.confirm ~default:false "Override?" then
+          (OpamConsole.msg
+             "You can revert this setting using \
+              'opam var --global os-distribution=%s'"
+             (OpamVariable.string_of_variable_contents vc);
+           update vars)
+        else
+          OpamStd.Sys.exit_because `Aborted
+    in
+    OpamFile.Config.with_sys_pkg_manager_cmd
+      (OpamStd.String.Map.add "cygwin" cygcheck
+         (OpamFile.Config.sys_pkg_manager_cmd config))
+      config
+  in
+  let get_cygwin = function
+    | Some cygcheck
+      when OpamFilename.exists cygcheck
+        && OpamStd.Sys.is_cygwin_cygcheck
+             ~cygbin:(Some OpamFilename.(Dir.to_string (dirname cygcheck))) ->
+      (* Should display the name using the converted path of /, not the bin dir *)
+      OpamConsole.note "Using Cygwin installation at %s for depexts"
+        (OpamFilename.(Dir.to_string (dirname cygcheck)));
+      success cygcheck
+    | Some _ | None ->
+      let rec menu () =
+        let enter_paths () =
+          let prompt_setup () =
+            let options = [
+              `manual, "Specify Cygwin setup path";
+              `download,
+              "Let opam downloads Cygwin setup and stores it in \
+               Cygwin installation path";
+              `abort, "Abort initialisation";
+            ]
+            in
+            OpamConsole.menu "Opam needs Cygwin setup executable" ~no:`download ~options
+          in
+          let rec enter_path () =
+            match OpamConsole.read "Enter the path to the Cygwin installation:" with
+            | None -> None
+            | Some "" ->
+              OpamConsole.msg "Empty path\n";
+              enter_path ()
```

Comment:
```suggestion
```
`OpamConsole.read` will return `None` and never `Some ""` on empty input, so this case seems superfluous. But since it's different from the previous one I don't know what to make of this :grimacing: 

Also be careful on looping on `None` since that's what non-interactive mode returns (but here it's guarded by the menu so should be OK)